packages/core/src/browser/preferences/preference-validation-service.ts

```diff
@@ -0,0 +1,280 @@
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { JsonType, PreferenceItem } from '../../common/preferences/preference-schema';
+import { JSONObject, JSONValue } from '../../../shared/@phosphor/coreutils';
+import { PreferenceSchemaProvider } from './preference-contribution';
+import { PreferenceLanguageOverrideService } from './preference-language-override-service';
+import { inject, injectable } from '../../../shared/inversify';
+import { IJSONSchema } from '../../common/json-schema';
+import { deepClone, unreachable } from '../../common';
+import { PreferenceProvider } from './preference-provider';
+
+export interface PreferenceValidator<T> {
+    name: string;
+    validate(value: unknown): T;
+}
+
+export interface ValidatablePreference extends IJSONSchema, Pick<PreferenceItem, 'defaultValue'> { }
+export type ValueValidator = (value: JSONValue) => JSONValue;
+
+export interface PreferenceValidationResult<T extends JSONValue> {
+    original: JSONValue | undefined;
+    valid: T;
+    messages: string[];
+}
+
+@injectable()
+export class PreferenceValidationService {
+    @inject(PreferenceSchemaProvider) protected readonly schemaProvider: PreferenceSchemaProvider;
+    @inject(PreferenceLanguageOverrideService) protected readonly languageOverrideService: PreferenceLanguageOverrideService;
+
+    validateOptions(options: Record<string, JSONValue>): Record<string, JSONValue> {
+        const valid: Record<string, JSONValue> = {};
+        let problemsDetected = false;
+        for (const [preferenceName, value] of Object.entries(options)) {
+            const validValue = this.validateByName(preferenceName, value);
+            if (validValue !== value) {
+                problemsDetected = true;
+            }
+            valid[preferenceName] = validValue;
+        }
+        return problemsDetected ? valid : options;
+    }
+
+    validateByName(preferenceName: string, value: JSONValue): JSONValue {
+        const schema = this.getSchema(preferenceName);
+        return this.validateBySchema(preferenceName, value, schema);
+    }
+
+    validateBySchema(key: string, value: JSONValue, schema: ValidatablePreference | undefined): JSONValue {
+        const validated = this.doValidateBySchema(key, value, schema);
+        if (validated !== value) {
+            console.warn(`While validating prefrences, found impermissible value for ${key}. Using valid value`, validated, 'instead of configured value', value);
```

Comment:
typo:

```suggestion
            console.warn(`While validating preferences, found impermissible value for ${key}. Using valid value`, validated, 'instead of configured value', value);
```